### PR TITLE
WIP: Test running mesos + kubernetes compute clusters

### PR DESCRIPTION
**Pull request by scrosby**
## Changes proposed in this PR

- Get our tests to pass with mesos or kubernetes within one scheduler invocation.

## Why are we making these changes?
Test kubernetes + mesos within one cook. This is planned as a throwaway PR, to test the integration test infrastructure to determine what changes are needed, there, to support kubernetes+mesos.

The good news: This version does pass integration testing, running locally (with integration testing on both k8s and mesos, run within a single launch.)

### CODE REVIEW

File: integration/tests/cook/test_basic.py [hunk 1]
```diff
@@ -720,7 +724,7 @@ def test_memory_limit_exceeded_mesos_script(self):
     def test_get_job(self):
         # schedule a job
         job_spec = util.minimal_job()
-        resp = util.session.post('%s/rawscheduler' % self.cook_url, json={'jobs': [job_spec]})
+        resp = util.session.post('%s/jobs' % self.cook_url, json={'jobs': [job_spec], 'pool' : util.default_submit_pool()})
```
Comment (dposada) [inline integration/tests/cook/test_basic.py]: We should just do:

```suggestion
        _, resp = util.submit_job(self.cook_url)
```

`submit_job` takes care of using the correct endpoint, submitting a `minimal_job()`, and using the `default_submit_pool()`.

File: integration/tests/cook/test_multi_user.py [hunk 2]
```diff
@@ -116,14 +116,14 @@ def test_job_cpu_quota(self):
...
-                resp = util.set_limit(self.cook_url, 'quota', user.name, cpus=0)
+                resp = util.set_limit(self.cook_url, 'quota', user.name, cpus=0, pool=util.default_submit_pool())
                 self.assertEqual(resp.status_code, 201, resp.text)
             with user:
                 _, resp = util.submit_job(self.cook_url)
                 self.assertEqual(resp.status_code, 422, msg=resp.text)
             # User with tiny quota can't submit bigger jobs, but can submit tiny jobs
             with admin:
-                resp = util.set_limit(self.cook_url, 'quota', user.name, cpus=0.25)
+                resp = util.set_limit(self.cook_url, 'quota', user.name, cpus=0.25, pool=util.default_submit_pool())
```
Comment (dposada) [inline integration/tests/cook/test_multi_user.py]: Seems like we should just make the default for all these limit-setting calls when no pool is specified be `default_submit_pool()`

File: integration/tests/cook/test_multi_user.py [hunk 3]
```diff
@@ -883,10 +883,10 @@ def test_queue_endpoint(self):
         job_spec = {'group': group['uuid'],
                     'command': 'sleep 30',
                     'cpus': util.max_cpus()}
-        uuids, resp = util.submit_jobs(self.cook_url, job_spec, clones=100, groups=[group])
+        uuids, resp = util.submit_jobs(self.cook_url, job_spec, clones=100, groups=[group], pool=util.default_submit_pool())
```
Comment (dposada) [inline integration/tests/cook/test_multi_user.py]: This is the default behavior

File: integration/tests/cook/util.py [hunk 4]
```diff
@@ -1635,16 +1640,17 @@ def _get_compute_cluster_factory_fn():
...
-    return _get_compute_cluster_factory_fn() == 'cook.mesos.mesos-compute-cluster/factory-fn'
-
+    return os.getenv("COOK_TEST_COMPUTE_CLUSTER_TYPE","mesos") == "mesos"
 
 def has_one_agent():
-    return node_count() == 1
-
+    if using_kubernetes():
+        # Kubernetes always has more than one agent.
+        return False
```
Comment (dposada) [inline integration/tests/cook/util.py]: Why?

File: integration/tests/cook/util.py [hunk 5]
```diff
@@ -1252,18 +1260,20 @@ def default_pool(cook_url):
 def all_pools(cook_url):
     """Returns the list of all pools that exist"""
     resp = session.get(f'{cook_url}/pools')
-    return resp.json(), resp
+    disallow_pools_regex = os.getenv('COOK_TEST_DISALLOW_POOLS_REGEX')
+    all_pools = resp.json() if disallow_pools_regex is None else \
+        [p for p in resp.json() if not re.match(disallow_pools_regex, p['name'])]
+    if disallow_pools_regex is not None:
+        for ii in resp.json():
+            logger.info(f"For Pool {ii['name']} we have {re.match(disallow_pools_regex, ii['name'])}")
```
Comment (dposada) [inline integration/tests/cook/util.py]: I would vote to either remove this log or make it a little more clear, e.g.:

```suggestion
            logger.info(f"For pool {ii['name']}, matching against the disallowed pools regex returns {re.match(disallow_pools_regex, ii['name'])}")
```

File: integration/tests/cook/util.py [hunk 6]
```diff
@@ -1635,12 +1645,11 @@ def _get_compute_cluster_factory_fn():
...
 
 def using_kubernetes():
-    return get_kubernetes_compute_cluster() is not None
+    return os.getenv("COOK_TEST_COMPUTE_CLUSTER_TYPE", "mesos") == "kubernetes"
 
 
 def using_mesos():
-    return _get_compute_cluster_factory_fn() == 'cook.mesos.mesos-compute-cluster/factory-fn'
-
+    return os.getenv("COOK_TEST_COMPUTE_CLUSTER_TYPE", "mesos") == "mesos"
```
Comment (dposada) [inline integration/tests/cook/util.py]: Seems like we should extract this into a function:

```
def compute_cluster_type_under_test():
    return os.getenv("COOK_TEST_COMPUTE_CLUSTER_TYPE", "mesos")
````

File: scheduler/Dockerfile [hunk 7]
```diff
@@ -45,8 +44,13 @@ COPY config* /opt/cook/
 COPY resources /opt/cook/resources
 COPY java /opt/cook/java
 COPY src /opt/cook/src
+
 RUN lein uberjar
 RUN cp "target/cook-$(lein print :version | tr -d '"').jar" datomic-free-0.9.5394/lib/cook-$(lein print :version | tr -d '"').jar
+COPY .cook_kubeconfig_* /opt/cook/
+
+# Ugly hack. Our .cook_kubeconfig lookup assumes it can be found in ../scheduler/ so make a symlink
+RUN ln -s /opt/cook /opt/scheduler
```
Comment (dposada) [inline scheduler/Dockerfile]: Aren't these the same two lines as above?

File: integration/tests/cook/conftest.py [hunk 8]
```diff
@@ -35,6 +35,7 @@ def _ssh_check(user):
     """
     hostname = socket.gethostname()
     logging.info(f'Checking ssh as {user} to {hostname}')
+    print(f'Checking ssh as {user} to {hostname}')
```
Comment (dposada) [inline integration/tests/cook/conftest.py]: Remove?

File: scheduler/bin/run-local-composite.sh [hunk 9]
```diff
@@ -0,0 +1,82 @@
+#!/usr/bin/env bash
+
+# Usage: ./bin/run-local.sh
```
Comment (dposada) [inline scheduler/bin/run-local-composite.sh]: This doesn't match the filename

File: scheduler/bin/run-local-composite.sh [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    keytool -genkeypair -keystore ${COOK_KEYSTORE_PATH} -storetype PKCS12 -storepass cookstore -dname "CN=cook, OU=Cook Developers, O=Two Sigma Investments, L=New York, ST=New York, C=US" -keyalg RSA -keysize 2048
+fi
+
+if ! [ -x "$(command -v flask)" ]; then
+    echo "Please install flask"
+    exit 1
+fi
+
+INTEGRATION_DIR="$(dirname ${SCHEDULER_DIR})/integration"
+FLASK_APP=${INTEGRATION_DIR}/src/data_locality/service.py flask run -p 35847 &
```
Comment (dposada) [inline scheduler/bin/run-local-composite.sh]: Can we ditch the data locality service? It's not used in any real environment.

File: scheduler/config-composite.edn [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+                              :config-file "../scheduler/.cook_kubeconfig_2"}}
+                    {:factory-fn cook.mesos.mesos-compute-cluster/factory-fn
+                     :config {:failover-timeout-ms nil ; When we close the instance of Cook, all its tasks are killed by Mesos
+                              :master #config/env "MESOS_MASTER"
+                              :framework-id #config/env "COOK_FRAMEWORK_ID"
+                              :compute-cluster-name "local-mesos"}}]
+ :cors-origins ["https?://cors.example.com"]
+ :data-local {:fitness-calculator {:cache-ttl-ms 60000
+                                   :cost-endpoint #config/env "DATA_LOCAL_ENDPOINT"
+                                   :update-interval-ms 1000}}
```
Comment (dposada) [inline scheduler/config-composite.edn]: Again, I would vote to ditch this.

File: integration/tests/cook/test_basic.py (hunk 1 above)
Comment (scrosby) [reply integration/tests/cook/test_basic.py]: Nice.

File: integration/tests/cook/test_multi_user.py (hunk 2 above)
Comment (scrosby) [reply integration/tests/cook/test_multi_user.py]: Done.

File: integration/tests/cook/test_multi_user.py (hunk 3 above)
Comment (scrosby) [reply integration/tests/cook/test_multi_user.py]: Yup. Nuked.

File: integration/tests/cook/util.py (hunk 4 above)
Comment (scrosby) [reply integration/tests/cook/util.py]: Simplifies this code. I hate that we run these tests in so many environments and randomly disable subsets of tests. No production environment will have only a single node.

File: integration/tests/cook/util.py (hunk 6 above)
Comment (scrosby) [reply integration/tests/cook/util.py]: Good idea.

File: scheduler/bin/run-local-composite.sh (hunk 9 above)
Comment (scrosby) [reply scheduler/bin/run-local-composite.sh]: FIxed.

File: scheduler/bin/run-local-composite.sh (hunk 10 above)
Comment (scrosby) [reply scheduler/bin/run-local-composite.sh]: Yes, but lets do it as another PR; maybe something for next week?

**Pull request closed without merging**